### ISSUE TITLE
Authentication for graphql schema

### ISSUE THREAD

**Issue by karaggeorge** [samdenty/gqless#25]
My graphql endpoint (using FuanaDB) requires authentication via a header. It would be nice if there was support for providing headers that are passed on to the endpoint, both on the querying side and the generation cli. 

**Workaround:** (at least for generating)

Setting up a temporary proxy to add the header to the request and using that to generate the schema

I'm probably going to work on a PR to add this in the next few days if it sounds reasonable

**Comment by samdenty** [comment]
That would be awesome!

Adding an extra option to the cli, and passing it through codegen [client.ts](https://github.com/samdenty/gqless/blob/master/packages/schema/src/Codegen/files/client.ts) should suffice

**Comment by karaggeorge** [comment]
I'll work on it this weekend. How do you envision it working on the client (react)? Since you import it and use it directly, there's no configuration. I was thinking optionally import it in a file, add some configuration and then export it with that config. So then instead of importing it from `@gqless/react` you import it from that file.

Thoughts?

**Comment by karaggeorge** [comment]
Also thinking for the CLI. Not sure how you would be able to pass generic headers through a CLI argument. Maybe make that focused on the `Authorization` header only? And then adapt only if there comes a need for more

**Comment by Yomanz** [comment]
It would need an option for what header to pass and where to get the actual information to pass from and idk where you'd get the latter.

**Comment by karaggeorge** [comment]
I was thinking the getAuth() function can be passed in, and then it can handle where it gets the token from. Looks like that's what apollo is doing, but they use the localStorage as a default method if you don't pass it in yourself. Can use a similar default in this case, since people might be used to that already

**Comment by samdenty** [comment]
I think replicating what apollo/apollo-tooling does, is a good option.

`gqless.config.js/yml/json` config file

Passing config via cli arguments eg. `--header`

**Comment by karaggeorge** [comment]
I have it working really well on the CLI side, with flags or a config file, very similar to Apollo (I used their implementation as an inspiration).

A couple of notes:

- I noticed that FaunDB also doesn't accept `GET` requests, so the schema introspect query fails. Should I add an option to allow `POST` requests for the generation?
- It adds the headers to the client correctly, but you might need to use different tokens there depending on the user logged in, and not one static one. And I would rather not keep it static in the code. Should we offer a way to pass headers when you instantiate the gqless client in the frontend? Or an option to pass that as a configuration during runtime (after login flow)

Pushing up a PR soon 😄
